### PR TITLE
Add support for async plugin cleanup function

### LINKED ISSUE #143: Add support for async plugin cleanup function
It is useful for testing to allow plugins to register cleanup functions. This will drastically simplify
testing any plugin that depends on side effects such as opening a long lived connection, timers, intervals, etc.

### PR DESCRIPTION
**Pull request by ganemone**
Fixes #143

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -196,6 +204,12 @@ export default createPlugin({
       return next();
     };
   },
+  // dependency injected async cleanup function
```
Comment (dennisgl) [inline README.md]: nit: "dependency injected" seems somewhat confusing to me given only `provided`(`thing`) is available

File: src/base-app.js [hunk 2]
```diff
@@ -49,6 +50,9 @@ class FusionApp {
     enhancers.push(enhancer);
     this.registered.set(getTokenRef(token), {value, aliases, enhancers});
   }
+  cleanup() {
+    return Promise.all(this.cleanups.map(fn => fn()));
```
Comment (lhorie) [inline src/base-app.js]: There should be also a way to query the status of an app, similar to `server.listening`

File: src/base-app.js [hunk 3]
```diff
@@ -118,6 +122,9 @@ class FusionApp {
       let provides = value;
       if (value && value.__plugin__) {
         provides = resolvePlugin(provides);
+        if (value.cleanup) {
```
Comment (lhorie) [inline src/base-app.js]: Loose check here allows type error to occur in line below

**Comment by ganemone** [comment]
!merge

File: src/base-app.js (hunk 2 above)
Comment (ganemone) [reply src/base-app.js]: If we end up needing that, it would be easy enough to add. I want to keep things as simple as possible for now though

**Pull request merged**